cmd/supernode/app/root.go

```diff
@@ -65,121 +122,208 @@ func setupFlags(cmd *cobra.Command, opt *Options) {
 	}
 
-	// initialize supernode logger.
-	if err := initLog(logrus.StandardLogger(), "app.log"); err != nil {
-		return err
+	for _, f := range flags {
+		if err := v.BindPFlag(f.key, rootCmd.Flag(f.flag)); err != nil {
+			return err
+		}
 	}
+	return nil
+}
 
-	// initialize dfget logger.
-	dfgetLogger := logrus.New()
-	if err := initLog(dfgetLogger, "dfget.log"); err != nil {
+// readConfigFile reads config file into the given viper instance. If we're
+// reading the default configuration file and the file does not exist, nil will
+// be returned.
+func readConfigFile(v *viper.Viper, cmd *cobra.Command) error {
+	v.SetConfigFile(v.GetString("config"))
+	v.SetConfigType("yaml")
+
+	if err := v.ReadInConfig(); err != nil {
+		// when the default config file is not found, ignore the error
+		if os.IsNotExist(err) && !cmd.Flag("config").Changed {
+			return nil
+		}
 		return err
 	}
 
-	if err := initConfig(); err != nil {
-		return err
-	}
+	return nil
+}
 
-	// set supernode advertise ip
-	if stringutils.IsEmptyStr(cfg.AdvertiseIP) {
-		if err := setAdvertiseIP(); err != nil {
-			return err
-		}
+// getConfigFromViper returns supernode config from the given viper instance
+func getConfigFromViper(v *viper.Viper) (*config.Config, error) {
+	cfg := config.NewConfig()
+
+	if err := v.Unmarshal(cfg, func(dc *mapstructure.DecoderConfig) {
+		dc.TagName = "yaml"
+		dc.DecodeHook = decodeWithYAML(
+			reflect.TypeOf(time.Second),
+			reflect.TypeOf(rate.B),
+			reflect.TypeOf(fileutils.B),
+		)
+	}); err != nil {
+		return nil, errors.Wrap(err, "unmarshal yaml")
 	}
-	logrus.Infof("success to init local ip of supernode, use ip: %s", cfg.AdvertiseIP)
 
-	// set up the CIDPrefix
-	cfg.SetCIDPrefix(cfg.AdvertiseIP)
+	// set dynamic configuration
+	cfg.DownloadPath = filepath.Join(cfg.HomeDir, "repo", "download")
 
-	logrus.Debugf("get supernode config: %+v", cfg)
-	logrus.Info("start to run supernode")
-
-	d, err := daemon.New(cfg, dfgetLogger)
-	if err != nil {
-		logrus.Errorf("failed to initialize daemon in supernode: %v", err)
-		return err
-	}
+	return cfg, nil
+}
 
-	// register supernode
-	if err := d.RegisterSuperNode(); err != nil {
-		logrus.Errorf("failed to register super node: %v", err)
-		return err
+// decodeWithYAML returns a mapstructure.DecodeHookFunc to decode the given
+// types by unmarshalling from yaml text.
+func decodeWithYAML(types ...reflect.Type) mapstructure.DecodeHookFunc {
+	return func(f, t reflect.Type, data interface{}) (interface{}, error) {
+		for _, typ := range types {
+			if t == typ {
+				b, _ := yaml.Marshal(data)
+				v := reflect.New(t)
+				return v.Interface(), yaml.Unmarshal(b, v.Interface())
+			}
+		}
+		return data, nil
 	}
-
-	return d.Run()
 }
 
 // initLog initializes log Level and log format.
 func initLog(logger *logrus.Logger, logPath string) error {
-	logFilePath := path.Join(options.HomeDir, "logs", logPath)
+	logFilePath := path.Join(supernodeViper.GetString("base.homeDir"), "logs", logPath)
```

Comment:
In fact, supernode will always use the default homeDir here because it's executed before reading config file. And #990 has tried to fix this. So I will merge this PR firstly.